File: src/services/tools/git/gitBranches.ts
import { z } from 'zod';
import type { ToolDef } from '../types';

export const gitBranchesTool: ToolDef = {
  name: 'git_branches',
  description: 'List all branches in the repository. Returns local and remote branches with their latest commit info.',
  category: 'git',
  inputSchema: z.object({
    repoPath: z.string().describe('Absolute path to the git repository root.'),
    remote: z.boolean().optional().default(false).describe('Include remote branches.'),
  }),
  execute: async ({ repoPath, remote }) => {
    const backendUrl = import.meta.env.VITE_API_URL || 'http://localhost:3001';
    try {
      const res = await fetch(`${backendUrl}/tools`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ tool: 'git_branches', params: { repoPath, remote } }),
      });
      if (!res.ok) throw new Error(`Backend error (${res.status})`);
      return res.json();
    } catch (error: any) {
      if (error.message?.includes('fetch')) {
        throw new Error('Git tools require the Express backend to be running (http://localhost:3001) or the Tauri desktop app');
      }
      throw error;
    }
  },
};

File: agent/internal/tool/git_branches.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func gitBranchesTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "git_branches",
			Description: "List all local and remote branches for a git repository.",
			Category:    "git",
			Parameters: map[string]api.ParamDef{
				"path":   {Type: "string", Description: "Path to the git repository", Required: false, Default: "."},
				"remote": {Type: "boolean", Description: "Include remote branches", Required: false, Default: true},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			path, _ := params["path"].(string)
			if path == "" {
				path = "."
			}
			args := []string{"branch"}
			if remote, ok := params["remote"].(bool); ok && remote {
				args = append(args, "-a")
			}
			out, err := gitCmd(ctx, path, args...)
			if err != nil {
				return nil, fmt.Errorf("git branches failed: %w", err)
			}
			return map[string]any{"path": path, "output": out}, nil
		},
	}
}
